# Bioinformatics Course for Beginners

Please follow the steps in the Preparation Page before the course starts.

A five weeks course for students and other life scientists who are just starting with bioinformatics.

Prerequisites for this course: Although not required, familiarity with python or a coding language is recommended. Participants should bring laptops.

## Objectives

We’ll start by exploring the most used bioinformatics resources for variety of different data sets, continue with practices on genomics, spatial transcriptomics and single cell sequencing. We’ll cover the best practices and common challenges in this practice as a foundation to the understanding of bioinformatics pipelines. Rest of the course, you’ll be working in smaller project groups. You’ll be able to choose projects on clustering, immunobiology, and cancer genomics, images etc. As you can see, majority of the course will be hands-on allowing you to practice your acquired knowledge.

At the end students will 

- realize the wealth of bioinformatics database structures and the -omic tools 

- understand how to download and use data from bioinformatics databases

- learn best practices and common challenges in genomics, single cell sequencing, spatial transcriptomics technologies

- explore downstream analyses approaches for bioinformatics datasets

- work on a project to practice acquired knowledge on a topic of interest

- work on datasets provided by real life science researchers on clustering, images, immunobiology, molecular biology, and cancer genomics

- present their findings 

## Content

### Week 1:

17:30 - 18:00 introductions

18:00 - 18:30 introduction to bioinformatics and databases (UCSC Genome browser, biomart, oma, string and others from SIB)

18:30 - 18:45 break

18:45 - 19:30 introduction to genomics

19:30 - 19:45 break

19:45 - 20:30 preparations for the genomics practice session

### Week 2:

17:30 - 18:00 introduction to HPC and terminal

18:00 - 18:30 setting up the workspace for the genomics practice

18:30 - 18:45 break

18:45 - 19:30 genomics practice 

19:30 - 19:45 break

19:45 - 20:30 discussion of the results from the practice session

### Week 2:

17:30 - 18:00 introduction to HPC and terminal

18:00 - 18:30 setting up the workspace for the genomics practice

18:30 - 18:45 break

18:45 - 19:30 genomics practice 

19:30 - 19:45 break

19:45 - 20:30 discussion of the results from the practice session




14:00 - 16:00 practice on genomics 

16:00 - 16:30 coffee break

16:30 - 17:00 project selection and meet up with project leaders

### Day 2:

09:00 - 09:30 Q&A, reflections

09:30 - 12:00 project work

12:00 - 13:00 lunch break

13:00 - 15:00 project work

15:00 - 16:30 presentations

16:30 - 17:00 wrap-up, conclusions, feedback

## Methodology

The module will consist of lectures and practical exercises. In addition to lectures, students will be required to self-study selected topics. Students will work in groups on a data challenge and present their results at the end of the course. 

Exercises during the course: 50%
Data challenge: 50%
The course is taught in English. We’ll be working in bash. You may also use python, R etc. depending on the projects. No prior knowledge in coding is required although familiarity with a coding language will be helpful. Please bring laptops. If you have questions, feel free to email Tugce.

## More details about the implementation

There are two days of classes on the last two Fridays of May. Some preliminary reading and preparation materials are provided. First day includes theory introduction and hands-on exercises. The second day includes practical guided work on the group projects, discussions and final presentations.

Copy the presentations from here:
```
cp /cfs/earth/scratch/icls/shared/bioinfo4beginners/'Practice Slides.pdf' .
```
