---
title: "Analysis, visualization, and integration of spatial datasets with Seurat"
output:
  html_document:
    theme: united
    df_print: kable
  pdf_document: default
date: 'Compiled: `r Sys.Date()`'
---

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  fig.width = 10,
  time_it = TRUE,
  error = TRUE
)
```

# Overview

This tutorial demonstrates how to use Seurat (>=3.2) to analyze spatially-resolved RNA-seq data. The analytical pipelines are similar to the Seurat workflow for [single-cell RNA-seq analysis](pbmc3k_tutorial.html), with a particular emphasis on the integration of spatial and molecular information. This tutorial will cover the following tasks:

* Normalization 
* Dimensional reduction and clustering
* Detecting spatially-variable features


We analyze a dataset generated with the [Visium technology](https://www.10xgenomics.com/spatial-transcriptomics/) from 10x Genomics. This notebook is modified from [this tutorial](https://satijalab.org/seurat/articles/spatial_vignette).

First, we load Seurat and the other packages necessary for this vignette.

```{r install}
devtools::install_github('satijalab/seurat-data')
library(Seurat)
library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)
```

```{r libraries.for.rmd, echo = FALSE}
library("htmltools")
library("vembedr")
```


# 10x Visium

## Dataset

Here, we will be using a recently released dataset of sagital mouse brain slices generated using the Visium v1 chemistry. There are two serial anterior sections, and two (matched) serial posterior sections. 

You can find this data [here](https://support.10xgenomics.com/spatial-gene-expression/datasets), and other outputs of the [spaceranger](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger) pipeline. The following code returns a Seurat object that contains both the spot-level expression data along with the associated image of the tissue slice. After installing the dataset, you can type `?stxBrain` to learn more.

```{r data.install, eval = FALSE}
InstallData("stxBrain")
```

```{r data}
brain <- LoadData('stxBrain', type = 'anterior1')
```
 
<details>
  <summary>**How is the spatial data stored within Seurat? **</summary>
The visium data from 10x consists of the following data types:

* A spot by gene expression matrix
* An image of the tissue slice (obtained from H&E staining during data acquisition)
* Scaling factors that relate the original high resolution image to the lower resolution image used here for visualization.         

In the Seurat object, the spot by gene expression matrix is similar to a typical "RNA" `Assay` but contains spot level, not single-cell level data. The image itself is stored in a new `images` slot in the Seurat object. The `images` slot also stores the information necessary to associate spots with their physical position on the tissue image.
</details>

## Data preprocessing

The initial preprocessing steps that we perform on the spot by gene expression data are similar to a typical scRNA-seq experiment. We first need to normalize the data in order to account for variance in sequencing depth across data points. We note that the variance in molecular counts / spot can be substantial for spatial datasets, particularly if there are differences in cell density across the tissue. We see substantial heterogeneity here, which requires effective normalization.

```{r qc, fig.height=5}
plot1 <- VlnPlot(brain, features = 'nCount_Spatial', pt.size = 0.1) + NoLegend()
plot2 <- SpatialFeaturePlot(brain, features = 'nCount_Spatial') + theme(legend.position = "right")
wrap_plots(plot1, plot2)
```

```{r preprocess}
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
```

These plots demonstrate that the variance in molecular counts across spots is not just technical in nature, but also is dependent on the tissue anatomy. For example, regions of the tissue that are depleted for neurons (such as the cortical white matter), reproducibly exhibit lower molecular counts. As a result, standard approaches (such as the `LogNormalize()` function), which force each data point to have the same underlying 'size' after normalization, can be problematic. 


## Gene expression visualization 

In Seurat, we have functionality to explore and interact with the inherently visual nature of spatial data. The `SpatialFeaturePlot()` function in Seurat extends `FeaturePlot()`, and can overlay molecular data on top of tissue histology. For example, in this data set of the mouse brain, the gene Hpca is a strong hippocampus marker and Ttr is a marker of the choroid plexus. 

```{r featureplot}
SpatialFeaturePlot(brain, features = c("Hpca", "Ttr"))
```

## Dimensionality reduction, clustering, and visualization

We can then proceed to run dimensionality reduction and clustering on the RNA expression data, using the same workflow as we use for scRNA-seq analysis.

```{r dim.cluster}
brain <- RunPCA(brain, assay = "SCT", verbose = FALSE)
brain <- FindNeighbors(brain, reduction = "pca", dims = 1:30)
brain <- FindClusters(brain, verbose = FALSE)
brain <- RunUMAP(brain, reduction = "pca", dims = 1:30)
```

We can then visualize the results of the clustering either in UMAP space (with `DimPlot()`). 

```{r dim.plots,fig.height=5}
DimPlot(brain, reduction = "umap", label = TRUE) 
```

You can also use the `cells.highlight` parameter to demarcate particular cells of interest on a `SpatialDimPlot()`. This can be very useful for distinguishing the spatial localization of individual clusters, as we show below:

```{r facetdim}
SpatialDimPlot(brain, cells.highlight = CellsByIdentities(object = brain,idents = c(2, 1, 4, 3, 5, 8)), facet.highlight = TRUE, ncol = 3)
```


## Identification of Spatially Variable Features

Seurat offers workflows to identify molecular features that correlate with spatial location within a tissue. You can perform differential expression based on pre-annotated anatomical regions within the tissue, which may be determined either from unsupervised clustering or prior knowledge. This strategy works in this case, as the clusters above exhibit clear spatial restriction.

```{r de, fig.height = 4}
de_markers <- FindMarkers(brain, ident.1 = 5, ident.2 = 6)
SpatialFeaturePlot(object = brain, features = rownames(de_markers)[1:3], alpha = c(0.1, 1), ncol = 3)
```

